Initial System.Net.WebSockets.Client implementation on Unix

stephentoub opened the pull request:
This PR provides an initial implementation of the client WebSockets library for Unix.  It's actually not Unix-specific, and rather is a managed implementation on top of System.Net.Sockets and System.Net.Security.

I opted to start with a managed implementation as there does not appear to be a universally-adopted native implementation we can utilize.  I looked to see if we could just adopt the Mono implementation, but it's based on HttpWebRequest, which in corefx is built on top of HttpClient, which isn't particularly amenable to being used for websockets (and libcurl does not have websockets support).

This is passing all of the unit/functional tests, including a few more I added, though I'm fairly confident that there are additional bugs lurking (i.e. I assume there's going to be a bug tail here).  I'm opting to check this in as an initial implementation now in hopes that others will be able to help with finding and fixing issues.  Along with that, I've spent close to zero time looking at performance; I'm sure there are opportunities to make meaningful improvements, but for now I've opted for simplicity and functionality.  There are also two features I didn't implement: proxies and credentials... but it doesn't look like either the WinHttp or WinRT implementations pay attention to those either.

cc: @ericeil, @cipop, @davidsh, @bartonjs
Fix https://github.com/dotnet/corefx/issues/2486

stephentoub commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                    _disposed = true;
+                    _keepAliveTimer.Dispose();
+                    _stream?.Dispose();
+                    _client?.Dispose();
```
I considered that, but opted not to as the main ClientWebSocket class does the inverse: Abort calls Dispose, but Dispose doesn't call Abort:
https://github.com/stephentoub/corefx/blob/b65762052589009332599af9d8a4585436f6d6fc/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs#L217-L242

bartonjs replied:
Fine by me :)

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                            uri.Host, 
+                            options.ClientCertificates, 
+                            SecurityProtocol.AllowedSecurityProtocols, 
+                            checkCertificateRevocation: false).ConfigureAwait(false);
```
Skipping revocation?

Is there no input host verification callback?

stephentoub replied:
> Skipping revocation?

I did that because:
- It's the default if you use one of the shorter overloads, and as a result it's what we do in other libs, like SqlClient.
- It's the default with HttpClient, where you only get revocation checking if you set CheckCertificateRevocationList to true.
- Etc.
I have no problem changing it, though, if you say that's the right thing to do.

> Is there no input host verification callback?

You mean a callback for verifying the remote certificate?  A RemoteCertificateValidationCallback delegate can be passed to the SslStream ctor to do manual validation of the server's certificate... but why do you ask?  Isn't the default verification sufficient?

bartonjs replied:
I guess my question really is "without looking at the input data types, are these options to SslStream currently exposed as options on the WebSocketRequest, so do they need to be passed along here?"

stephentoub replied:
> are these options to SslStream currently exposed as options on the WebSocketRequest

They are not; the only related property is ClientCertificates, which is being passed in here.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
-        }
+            public override Task CloseAsync(WebSocketCloseStatus closeStatus, string statusDescription, CancellationToken cancellationToken)
+            {
+                ClientWebSocket.ThrowIfInvalidState(_state, _disposed, s_validCloseStates);
```
Hard throw, not Task.FromException?  (also applies to CloseOutputAsync)

davidsh replied:
Consider #4508.

stephentoub replied:
I'll change them all to be in tasks as long as it doesn't break with existing tests, e.g. the Windows implementation is inconsistent here:
https://github.com/stephentoub/corefx/blob/b65762052589009332599af9d8a4585436f6d6fc/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs#L527-L532

davidsh replied:
>I'll change them all to be in tasks as long as it doesn't break with existing tests, e.g. the Windows implementation is inconsistent here:
https://github.com/stephentoub/corefx/blob/b65762052589009332599af9d8a4585436f6d6fc/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs#L527-L532

You should not just change them all to be tasks without considering #4508.

You should also look at Desktop implementation and test what how it behaves as well

stephentoub replied:
> You should not just change them all to be tasks without considering #4508.

I won't... I was the one who initially created the guideline being quoted in #4508 ;)

But everything in this file is only hit after it's already gone through the validation done by the public ClientWebSocket surface area... pretty much everything here isn't based purely on doing argument validation.  This case, for example, is based on the state of the instance, not just on the arguments, and in part based on race conditions... these make sense to be in the task.

davidsh replied:
ok. sounds like a good plan.

bartonjs replied:
I'm fine either way, I just noticed "two of these things don't look like the others" :)

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                        string header = options.Cookies.GetCookieHeader(uri);
+                        if (!string.IsNullOrWhiteSpace(header))
+                        {
+                            builder.Append(HttpKnownHeaderNames.SetCookie).Append(": ").Append(header).Append("\r\n");
```
Do we want to do proactive (or even reactive) header chopping for HTTP 413 protection?  (A quick search says the default for Apache is/was 8k per header, and IIS is/was 16k; but those are/were defaults).

So something like  if (header.Length > 8100) { find a semicolon before that mark, send up to there, newline; repeat ad infinitum }

stephentoub replied:
I looked in Wireshark at what happens when super long cookies are sent with ClientWebSocket in both the desktop and with corefx (e.g. 5 cookies each 4096 bytes long), and I don't see any such thing being done... there's just a single Set-Cookie header with all of the cookies as part of it.  My understanding of header limits is typically, if a server does impose one even though HTTP doesn't have an explicit limit, that the limit typically applies to all headers, not to a single header, so chopping into multiple Set-Cookie headers wouldn't help.

bartonjs replied:
Okay, thanks for checking.

Tratcher replied:
Umm, Set-Cookie is for Server->Client. Cookie is for Client->Server.

stephentoub replied:
Will fix

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                                SR.Format(SR.net_WebSockets_AcceptUnsupportedProtocol, string.Join(", ", options.RequestedSubProtocols), subprotocol));
+                        }
+                        subprotocol = newSubprotocol;
+                    }
```
Do we not need to save other headers for anything else?  Since we were reading here it seems like any data we don't save is just gone.

stephentoub replied:
What else would we need to save?  To my knowledge based on RFC 6455, the answer is "no".

bartonjs replied:
Fair enough; I hadn't opened the RFC at all until I got to the header framing part :)

stephentoub commented:
Test Innerloop Windows_NT Release Build and Test please (https://github.com/dotnet/corefx/issues/8926)

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                {
+                    if (isTargetHeader)
+                    {
+                        throw new WebSocketException(SR.Format(SR.net_webstatus_ConnectFailure));
```
I notice that there are a lot of places that can cause ConnectFailure; and they're potentially hard to diagnose.  Do we need/want logging in this class?

stephentoub replied:
> Do we need/want logging in this class?

I want to add detailed EventSource-based logging, similar to what we have in CurlHandler, but we'll do that separately.  We don't have a WebSockets-specific EventSource, and I'd like to add one subsequently.

bartonjs replied:
The logging plan sounds good to me.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+            private static KeyValuePair<string, string> CreateSecKeyAndSecWebSocketAccept()
+            {
+                string secKey = Convert.ToBase64String(Guid.NewGuid().ToByteArray());
+                using (SHA1 sha = SHA1.Create())
```
I expect that you considered saving a SHA1 instance into a (non-thread) static field and using Interlocked.Exchange to round-robin it (and Dispose of any parallel excesses).

On the off-chance that you didn't, consider it offered.
If you did, and didn't use it: I don't disagree. (Since I don't know the expected parallel factor of this code)

Or there's always instance method and make it an instance field, so it tears down sympathetically.

stephentoub replied:
I didn't consider it much.  Are they very expensive?  It's currently one created and used briefly per WebSocket.  If they're hefty, I can easily follow your suggestion.

bartonjs replied:
It's not terribly expensive; I only brought it up because of the existing caching.  I'd, personally, leave this as-is unless data drove me to do otherwise.

stephentoub replied:
Honestly, there are much more expensive things happening elsewhere in the implementation :wink:  I'll leave this as-is for now.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                    {
+                        int numRead = await stream.ReadAsync(oneByte, 0, 1, cancellationToken).ConfigureAwait(false);
+                        byte b = oneByte[0];
+                        if (numRead == 0 || b == '\n') break;
```
Since the spec says that newlines are CRLF, can an LF appear within a header?  Do we need to track the two-character state?

stephentoub replied:
I don't believe that's allowed, but I can change this to check the previous character regardless.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+            private void SendKeepAliveFrameAsync()
+            {
+                Task pingTask = SendFrameAsync(new ArraySegment<byte>(Array.Empty<byte>()), MessageOpcode.Ping, true, CancellationToken.None);
+                // This is exists purely to keep the connection alive; ignore any failures.
```
Do you not need to do something like GetAwaiter().GetResult() to make it complete?

stephentoub replied:
I explicitly don't want to (that's what the comment was trying to get at, but I guess I didn't explain enough).  I simply want to initiate the operation, and there's no need to block this thread waiting for it to complete. The only reason to do so would be if we wanted to propagate any failures on this thread, but I want to ignore any failures.

bartonjs replied:
Okay, I'll completely defer to you here.  I just wanted to make sure that we did enough work to pulse the keepalive across the socket, and not just got ready to do so once an awaiter was pumped.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                    for (int i = 9; i >= 2; i--)
+                    {
+                        sendBuffer[i] = (byte)(length % 256);
+                        length = length / 256;
```
Isn't `(byte)(length % 256)` just a more expensive way of writing unchecked((byte)length)?

Presumably the JIT sees length / 256 as length >> 8.

stephentoub replied:
Yeah, the mod is unnecessary, and it does result in worse code.  I'll fix it.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                //     - For length 0 through 125, 7 bits storing the length
+                //     - For lengths 126 through 2^16, 7 bits storing the value 126, followed by 16 bits storing the length
+                //     - For lengths 2^16+1 through 2^64, 7 bits storing the value 127, followed by 64 bytes storing the length
+                // 4 bits - Mask - 32-bit random value XOR'd with each 4 bytes of the payload, round-robin
```
Reading this as written here this would be misaligned data.  I think you mean 4 BYTES / 32 bits? :smile:

stephentoub replied:
Oops, yes, 4 bytes, will fix the comment.

bartonjs commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+                bool masked = (_receiveHeaderBuffer[1] & 0x80) != 0;
+                if (masked)
+                {
+                    // The server must not masked responses.  If it does, the client must close the connection.
```
Nit: `masked` => `mask`

stephentoub replied:
I need a comment grammar checker extension for VS :)

bartonjs commented on src/System.Net.WebSockets.Client/tests/ClientWebSocketTest.cs:
```diff
@@ -372,8 +412,73 @@ public static IEnumerable<object[]> UnavailableWebSocketServers
+                var ctsDefault = new CancellationTokenSource(TimeOutMilliseconds);
+
+                // Values chosen close to boundaries in websockets message length handling.
+                foreach (int bufferSize in new int[] { 1, 125, 126, 127, 128, ushort.MaxValue - 1, ushort.MaxValue, ushort.MaxValue + 1, ushort.MaxValue * 2 }) 
```
Is 0 legal? Interesting?

stephentoub replied:
We have tests elsewhere for 0.  I don't think it's particularly interesting for what this test is trying to handle.

bartonjs commented on src/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:
```diff
@@ -64,21 +64,28 @@ public static class WebSocketHelper
             int timeOutMilliseconds,
-            ITestOutputHelper output)
+            ITestOutputHelper output,
+            TimeSpan keepAliveInternal = default(TimeSpan))
```
InterNal => InterVal; yes? :)

stephentoub replied:
:smile:

davidfowl commented:
/cc @Tratcher

Tratcher commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
-            throw new PlatformNotSupportedException(SR.net_WebSockets_UnsupportedPlatform);
-        }
+                    // Add all of the required headers
+                    builder.Append("Host: ").Append(uri.Host).Append("\r\n");
```
Host needs to be pynycode, and you need to include the port.

stephentoub replied:
Will change to IdnHost

Tratcher commented:
Were you able to re-use any of the code from here: https://github.com/aspnet/WebSockets/tree/dev/src/Microsoft.AspNetCore.WebSockets.Protocol or https://github.com/aspnet/WebSockets/blob/dev/src/Microsoft.AspNetCore.WebSockets.Client/WebSocketClient.cs

If this is going to stay managed then we should de-duplicate some code.

Have you run the standard WebSocket autobahn tests? http://autobahn.ws/. There's a setup for it here: https://github.com/aspnet/WebSockets/tree/dev/test

ericeil commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs:
```diff
@@ -2,108 +2,1037 @@
+            public ManagedClientWebSocket()
             {
-                throw new PlatformNotSupportedException(SR.net_WebSockets_UnsupportedPlatform);
+                // Initialize the keep alive timer.  When active, it'll send periodic ping frames to the server.
```
How critical is the timing of this?  What if a ping is delayed for a very long time?  Given that Timers might be delayed by other items in the ThreadPool, should we consider a more dependable time source?

davidsh replied:
>How critical is the timing of this? 

RFC6455 talks about ping/pong and unsolicited pong frames.

stephentoub replied:
It's not particularly critical.  Common values might be something like 30 seconds, and the rfc highlights that things are pretty lenient.

stephentoub commented:
> https://github.com/aspnet/WebSockets/blob/dev/src/Microsoft.AspNetCore.WebSockets.Client/WebSocketClient.cs

I'd looked at that, but basing this on HttpWebRequest when HttpWebRequest is based on HttpClient won't work well.

> https://github.com/aspnet/WebSockets/tree/dev/src/Microsoft.AspNetCore.WebSockets.Protocol

Huh, I'd looked around, but I actually didn't see that one.  Would have saved me a day.  Skimming through it, I see value in pieces from both.  After getting this merged so that something is in place, I'll take a look at either adopting this instead or taking relevant pieces from it.

> Have you run the standard WebSocket autobahn tests? http://autobahn.ws/. There's a setup for it here: https://github.com/aspnet/WebSockets/tree/dev/test

Thanks for the link.  I haven't but I was interested.  I'll give it a shot.

Tratcher commented:
@stephentoub I wasn't expecting you to use the HttpWebRequest parts, but some of the header setup and validation may be useful.

stephentoub commented:
> Have you run the standard WebSocket autobahn tests?

Thanks for the pointer to the tests.  I addressed the feedback, made a few fixes (including using the UTF8 validation from ASP.NET), and it's now passing all of the Autobahn tests (except for the "Unimplemented" compression ones), though two of them show up as passing but "non-strict"... I need to look into why, but I'll do that separately.

Tratcher commented:
Yeah, there were one or two that seemed overly strict. Something about requiring you to abort on the first invalid UTF-8 byte rather than waiting for the full multi-byte sequence before validating.

futzy314 commented:
Has this fix made it into a release yet?  Or do we need to be running off of some dev build to see this fix?

stephentoub commented:
>Has this fix made it into a release yet?

It's in master and should be in the 1.1 release, whenever that ships.  If you want daily builds, you can use the ones from https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Net.WebSockets.Client

futzy314 commented:
Thanks, that was exactly what I needed and it is working perfectly.

stephentoub commented:
> Thanks, that was exactly what I needed and it is working perfectly.

Great to hear.  Thanks for letting us know.

FlorianHockmann commented:
Could you please clarify whether this was actually included in the 1.1 release?

As I still get a `PlatformNotSupportedException` on my Ubuntu machine that says _The WebSocket protocol is not supported on this platform._ when I call `ConnectAsync` on `ClientWebSocket` although I have installed version 1.1.

mazong1123 commented:
>  It's actually not Unix-specific, and rather is a managed implementation on top of System.Net.Sockets and System.Net.Security.

Seems like the implementation does not support in Windows 7. It gave me `Unhandled Exception: System.PlatformNotSupportedException: The WebSocket protocol is not supported on this platform.` Do we have any plan to support Windows 7 as well?

davidfowl commented:
@stephentoub you know what would be awesome? If we supported passing an HttpMessageHandler to the ClientWebSocket. It would be applied for the initial upgrade request. That would let me write the same logic for websockets as well as non websockets. (This is hitting us now with the SignalR client)
